File: typescript/perfect-numbers/perfect-numbers.ts
export default class PerfectNumbers {

    static classify(input: number): string {
        if (input <= 0) {
            throw new Error('Classification is only possible for natural numbers.')
        }
        const factors: number[] = this.getFactorsOf(input)
        const sum = factors.reduce((a, b) => a + b, 0)
        let res: string = ''
        sum === input ? res = 'perfect' :
            sum > input ? res = 'abundant' :
                sum < input ? res = 'deficient' :
                    false
        return res
    }

    static getFactorsOf(input: number): number[] {
        let factors: number[] = []
        for (let i = 1; i < input; i++) {
            if (Number.isSafeInteger(input / i) && i !== input) {
                factors.push(i)
            }
        }
        return factors
    }

}

File: csharp/perfect-numbers/PerfectNumbers.cs
using System;

public enum Classification
{
    Perfect,
    Abundant,
    Deficient
}

public static class PerfectNumbers
{
    public static Classification Classify(int number)
    {
        int minimumNumber = 1;
        int sumOfFactors = 0;
        Classification classification = new Classification();

        if (number < minimumNumber)
        {
            throw new ArgumentOutOfRangeException();
        }

        for (int i = minimumNumber; i < number - 1; i++)
        {
            if (number % i == 0)
            {
                sumOfFactors += i;
            }
        }

        if (sumOfFactors < number)
        {
            classification = Classification.Deficient;
        }
        else if (sumOfFactors == number)
        {
            classification = Classification.Perfect;
        }
        else if (sumOfFactors > number)
        {
            classification = Classification.Abundant;
        }

        return classification;
    }
}
